### PR TITLE
Add Calyx flow to `hlstool`

### LINKED ISSUE #5844: [Calyx] Add Calyx lowering flow to `hlstool`
As mentioned in #5544, we should add the Calyx lowering pipeline to `hlstool` so people have a good reference for where to get started.

### PR DESCRIPTION
**Pull request by rachitnigam**
Fixes #5844. Attempting to add a simple `scf` to `hw` lowering flow using `calyx` to `hlstool`. Once this lands, I can work on adding the `affine` to `loopschedule` to `calyx` flow as well.

### CODE REVIEW

**Comment by rachitnigam** [comment]
@mortbopet I might have a couple of dumb questions when working on this. The first one is that when I attempt to run this:
```
$ ./bin/hlstool --calyx-hw ../integration_test/Dialect/Calyx/simple_arith.mlir

../integration_test/Dialect/Calyx/simple_arith.mlir:0:0: error: 'builtin.module' op trying to schedule a pass on an unsupported operation
../integration_test/Dialect/Calyx/simple_arith.mlir:0:0: note: see current operation: 
"builtin.module"() ({
  "calyx.component"() ({
  ^bb0(%arg0: i32, %arg1: i32, %arg2: i1, %arg3: i1, %arg4: i1, %arg5: i32, %arg6: i1):
    %0:6 = "calyx.register"() {sym_name = "std_slt_0_reg"} : () -> (i1, i1, i1, i1, i1, i1)
...
```

I'm not sure what exactly doing wrong here (and don't have the right tools to look into which part of the pass pipeline causes this error). Do you have any pointers for me on how to debug this?

**Comment by rachitnigam** [comment]
Okay, I figure it out! It's a bunch of nesting of passes on particular ops. Not sure if there is a more elegant way of doing it but I'll leave that question to the review.

File: tools/hlstool/hlstool.cpp [hunk 1]
```diff
@@ -349,6 +360,61 @@ static LogicalResult doHLSFlowDynamic(
...
+
+  if (irInputLevel < 0) {
+    irInputLevel = HLSFlowDynamicIRLevel::High; // Default to highest level
+  }
+  if (irOutputLevel < 0) {
+    irOutputLevel = HLSFlowDynamicIRLevel::Sv; // Default to lowest level
+  }
+
+  if (irInputLevel != HLSFlowDynamicIRLevel::High) {
+    llvm::errs() << "Error: Calyx flow only supports 'high' IR input\n";
```
Comment (cgyurgyik) [inline tools/hlstool/hlstool.cpp:375]: Nit: (here and above) this "Error:" in the beginning of the string seems redundant.

**Review by cgyurgyik — APPROVED** [review]

File: tools/hlstool/hlstool.cpp (hunk 1 above)
Comment (rachitnigam) [reply tools/hlstool/hlstool.cpp:375]: Thanks, fixed!

File: integration_test/Dialect/Calyx/memory_loop.mlir [hunk 2]
```diff
@@ -1,11 +1,6 @@
 // This test lowers an SCF construct through Calyx, FSM and (TODO)
 // to RTL.
-// RUN: circt-opt %s \
-// RUN:     --lower-scf-to-calyx -canonicalize \
-// RUN:     --calyx-remove-comb-groups --canonicalize \
-// RUN:     --lower-calyx-to-fsm --canonicalize \
-// RUN:     --materialize-calyx-to-fsm --canonicalize \
-// RUN:     --calyx-remove-groups-fsm --canonicalize | FileCheck %s
+// RUN: hls-tool --calyx-hw --ir-output-level=2 %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply integration_test/Dialect/Calyx/memory_loop.mlir]: Fully agree - it was just something i added when i developed it. Probably the most sane is for each flow to define its own set of textual "checkpoints" that the user can specify.

**Review by mortbopet — APPROVED** [review]
LGTM once tests pass. We should also add a documentation page about HLS in CIRCT, i.e. at https://circt.llvm.org/docs/ - describing the high-level flows, hlstools, + further links to more in-depth information about each HLS type (probably just links to the dialect rationales).

**Review by mikeurbach — APPROVED** [review]
Awesome, thanks @rachitnigam! Having the flow stitched up in hlstool is great.

**Comment by rachitnigam** [comment]
Okay, I’m going to merge this and we can starting working on the follow up!

**Pull request merged**